### PR TITLE
[sk] Added multithreaded column type inference

### PR DESCRIPTION
**Pull request by skunichetty**
# Summary
Added support for column type inference to use multiple threads. When there are above a certain number of entries, the column type detection strategy is made concurrent by spawning multiple threads to perform calculation. This is done because if there are fewer entries, the overhead of multithreading outweighs the speed of the column type inference.

In additional, multicurrency support was added to column type inference, along with related bug fixes.

# Tests
Unit tests were ran and testing was also performed on testing datasets.

cc:
@wangxiaoyou1993

### CODE REVIEW

**Comment by wangxiaoyou1993** [comment]
Could you add performance improvement metrics to the PR description?

**Comment by skunichetty** [comment]
Added Performance Metrics

File: mage_ai/data_cleaner/column_type_detector.py [hunk 1]
```diff
@@ -54,143 +56,117 @@ def get_mismatched_row_count(series, column_type):
...
-    return column_types
+def infer_column_types(df, **kwargs):
+    columns = [df[col] for col in df.columns]
+    column_names = list(df.columns)
+    dtypes = list(df.dtypes)
+    kwarg_list = [kwargs] * len(columns)
+    ctypes = {}
+    num_entries = len(df)
+    if num_entries > MULTITHREAD_MAX_NUM_ENTRIES:
+        with ThreadPoolExecutor(max_workers=MAX_WORKERS) as pool:
```
Comment (wangxiaoyou1993) [inline mage_ai/data_cleaner/column_type_detector.py:166]: Let's add the multi thread method to https://github.com/mage-ai/mage-ai/blob/master/mage_ai/data_cleaner/shared/multi.py and use here

File: mage_ai/data_cleaner/column_type_detector.py [hunk 2]
```diff
@@ -35,14 +35,16 @@
     ZIP_CODE,
 ])
 
-REGEX_DATETIME_PATTERN = r'^[\d]{2,4}-[\d]{1,2}-[\d]{1,2}$|^[\d]{2,4}-[\d]{1,2}-[\d]{1,2}[Tt ]{1}[\d]{1,2}:[\d]{1,2}[:]{0,1}[\d]{1,2}[\.]{0,1}[\d]*|^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z$|^\d{1,4}[-\/]{1}\d{1,2}[-\/]{1}\d{1,4}$|(Jan(?:uary)?|Feb(?:ruary)?|Mar(?:ch)?|Apr(?:il)?|May|Jun(?:e)?|Jul(?:y)?|Aug(?:ust)?|Sep(?:tember)?|Oct(?:ober)?|Nov(?:ember)?|Dec(?:ember)?)\s+(\d{1,2})[\s,]+(\d{2,4})'
-REGEX_EMAIL_PATTERN = r"^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$"
+REGEX_DATETIME_PATTERN = r'^\d{2,4}-\d{1,2}-\d{1,2}$|^\d{2,4}-\d{1,2}-\d{1,2}[Tt ]{1}\d{1,2}:\d{1,2}[:]{0,1}\d{1,2}[\.]{0,1}\d*|^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z$|^\d{1,4}[-\/]{1}\d{1,2}[-\/]{1}\d{1,4}$|(Jan(?:uary)?|Feb(?:ruary)?|Mar(?:ch)?|Apr(?:il)?|May|Jun(?:e)?|Jul(?:y)?|Aug(?:ust)?|Sep(?:tember)?|Oct(?:ober)?|Nov(?:ember)?|Dec(?:ember)?)\s+(\d{1,2})[\s,]+(\d{2,4})'
```
Comment (wangxiaoyou1993) [inline mage_ai/data_cleaner/column_type_detector.py:38]: Could you summarize what updates are made to these regex?

File: mage_ai/data_cleaner/column_type_detector.py [hunk 3]
```diff
@@ -54,143 +56,117 @@ def get_mismatched_row_count(series, column_type):
...
+                    elif clean_series_unique / length >= 0.8:
+                        mdtype = TEXT
+                    else:
+                        word_count = clean_series.map(lambda x: len(str(x).split(' '))).max()
+                        if word_count > MAXIMUM_WORD_LENGTH_FOR_CATEGORY_FEATURES:
+                            mdtype = TEXT
+                        else:
+                            if (clean_series_unique <= kwargs.get(
+                                'category_cardinality_threshold', 255)):
+                                mdtype =  CATEGORY
```
Comment (wangxiaoyou1993) [inline mage_ai/data_cleaner/column_type_detector.py:153]: nit: remove one additional space after =

File: mage_ai/data_cleaner/column_type_detector.py (hunk 2 above)
Comment (skunichetty) [reply mage_ai/data_cleaner/column_type_detector.py:39]: Updated. A couple of the changes I made to these regex strings are redundant so I reverted them in the latest push.

File: mage_ai/data_cleaner/column_type_detector.py (hunk 1 above)
Comment (skunichetty) [reply mage_ai/data_cleaner/column_type_detector.py]: Added

File: mage_ai/data_cleaner/column_type_detector.py [hunk 4]
```diff
@@ -54,143 +57,114 @@ def get_mismatched_row_count(series, column_type):
...
         str_series = series.astype(str)
         mismatched_rows = len(
-            series[~str_series.str.contains(REGEX_ZIP_CODE)].index,
+            series[~str_series.str.match(REGEX_ZIP_CODE)].index,
         )
     return mismatched_rows
 
+def clean_whitespace(series):
+    return series.map(
+        lambda x: x if (not isinstance(x, str) or x != '') else np.nan,
```
Comment (wangxiaoyou1993) [inline mage_ai/data_cleaner/column_type_detector.py:75]: This doesn't clean string with multiple spaces, right?

File: mage_ai/data_cleaner/column_type_detector.py [hunk 5]
```diff
@@ -54,143 +57,114 @@ def get_mismatched_row_count(series, column_type):
...
         )
     elif column_type == ZIP_CODE:
         str_series = series.astype(str)
         mismatched_rows = len(
-            series[~str_series.str.contains(REGEX_ZIP_CODE)].index,
+            series[~str_series.str.match(REGEX_ZIP_CODE)].index,
         )
     return mismatched_rows
 
+def clean_whitespace(series):
```
Comment (wangxiaoyou1993) [inline mage_ai/data_cleaner/column_type_detector.py:73]: Is this method used in multiple places?

File: mage_ai/data_cleaner/column_type_detector.py [hunk 6]
```diff
@@ -54,143 +57,114 @@ def get_mismatched_row_count(series, column_type):
...
 
-    number_feature_names = float_feature_names + integer_feature_names
-    binary_feature_names += \
-        [col for col in number_feature_names if df[col].nunique(dropna=False) == 2]
-    binary_feature_names += \
-        [col for col in non_number_feature_names if df[col].nunique(dropna=False) == 2]
-    float_feature_names = [col for col in float_feature_names if col not in binary_feature_names]
-    integer_feature_names = \
-        [col for col in integer_feature_names if col not in binary_feature_names]
+    series_unique = series.nunique(dropna=False)
```
Comment (wangxiaoyou1993) [inline mage_ai/data_cleaner/column_type_detector.py:106]: nit: series_nunique and clean_series_nunique

File: mage_ai/data_cleaner/column_type_detector.py [hunk 7]
```diff
@@ -54,143 +57,114 @@ def get_mismatched_row_count(series, column_type):
...
+    clean_series_unique = clean_series.nunique()
 
-    for col_name in subtract(non_number_feature_names, binary_feature_names):
-        df_drop_na = df[col_name].dropna()
-        if df_drop_na.empty:
-            text_feature_names.append(col_name)
+    exact_dtype = type(clean_series.iloc[0])
```
Comment (wangxiaoyou1993) [inline mage_ai/data_cleaner/column_type_detector.py:109]: there will be error if clean_series is empty

File: mage_ai/data_cleaner/column_type_detector.py [hunk 8]
```diff
@@ -54,143 +57,114 @@ def get_mismatched_row_count(series, column_type):
...
             else:
-                word_count, _ = \
-                    df[col_name].dropna().map(lambda x: (len(str(x).split(' ')), str(x))).max()
-                if word_count > MAXIMUM_WORD_LENGTH_FOR_CATEGORY_FEATURES:
-                    text_feature_names.append(col_name)
+                mdtype = NUMBER
+        else:
+            length = len(clean_series)
+            matches = clean_series.str.match(REGEX_DATETIME).sum()
+            if matches/length >= DATETIME_MATCHES_THRESHOLD:
```
Comment (wangxiaoyou1993) [inline mage_ai/data_cleaner/column_type_detector.py:128]: nit: put space before and after '/'

File: mage_ai/data_cleaner/column_type_detector.py [hunk 9]
```diff
@@ -54,143 +57,114 @@ def get_mismatched_row_count(series, column_type):
...
+                mdtype = DATETIME
+            else:
+                correct_emails = clean_series.str.match(REGEX_EMAIL).sum()
+                correct_phone_nums = clean_series.str.match(REGEX_PHONE_NUMBER).sum()
+                correct_zip_codes = clean_series.str.match(REGEX_ZIP_CODE).sum()
+                if correct_emails/length >= 0.90:
```
Comment (wangxiaoyou1993) [inline mage_ai/data_cleaner/column_type_detector.py:134]: nit: space before and after '/'. same for correct_phone_nums and correct_zip_codes below

File: mage_ai/data_cleaner/shared/array.py [hunk 10]
```diff
@@ -28,4 +29,5 @@ def sample(arr):
 
 
 def subtract(arr1, arr2):
-    return [i for i in arr1 if i not in arr2]
+    arr2_lookup = set(arr2)
+    yield from (i for i in arr1 if i not in arr2_lookup)
```
Comment (wangxiaoyou1993) [inline mage_ai/data_cleaner/shared/array.py:33]: why do you use `yield from` here? it changes the behavior of this method.

File: mage_ai/data_cleaner/shared/utils.py [hunk 11]
```diff
@@ -19,21 +20,22 @@ def clean_series(series, column_type, dropna=True):
...
     
     first_item = series_cleaned.dropna().iloc[0]
-    if column_type == NUMBER:
+    if column_type == NUMBER or column_type == NUMBER_WITH_DECIMALS:
+        is_percent = False
         if type(first_item) is str:
             series_cleaned = series_cleaned.str.replace(',', '')
-            if series_cleaned.str.contains(CURRENCY_SYMBOLS).sum() != 0:
-                return # handle this case with reformat transform action
+            if series_cleaned.str.count(CURRENCY_SYMBOLS).sum() != 0:
```
Comment (wangxiaoyou1993) [inline mage_ai/data_cleaner/shared/utils.py:27]: Does the reformat_values cleaning rule still work?

File: mage_ai/data_cleaner/column_type_detector.py (hunk 5 above)
Comment (skunichetty) [reply mage_ai/data_cleaner/column_type_detector.py]: It is not used in multiple places, this function is now inlined.

File: mage_ai/data_cleaner/column_type_detector.py (hunk 6 above)
Comment (skunichetty) [reply mage_ai/data_cleaner/column_type_detector.py]: Fixed

File: mage_ai/data_cleaner/column_type_detector.py (hunk 8 above)
Comment (skunichetty) [reply mage_ai/data_cleaner/column_type_detector.py]: Fixed

File: mage_ai/data_cleaner/column_type_detector.py (hunk 9 above)
Comment (skunichetty) [reply mage_ai/data_cleaner/column_type_detector.py]: Fixed

File: mage_ai/data_cleaner/column_type_detector.py [hunk 12]
```diff
@@ -54,143 +56,117 @@ def get_mismatched_row_count(series, column_type):
...
+                mdtype = DATETIME
             else:
-                word_count, _ = \
-                    df[col_name].dropna().map(lambda x: (len(str(x).split(' ')), str(x))).max()
-                if word_count > MAXIMUM_WORD_LENGTH_FOR_CATEGORY_FEATURES:
-                    text_feature_names.append(col_name)
+                correct_emails = clean_series.str.match(REGEX_EMAIL).sum()
+                correct_phone_nums = clean_series.str.match(REGEX_PHONE_NUMBER).sum()
+                correct_zip_codes = clean_series.str.match(REGEX_ZIP_CODE).sum()
+                if correct_emails / length >= 0.90:
```
Comment (wangxiaoyou1993) [inline mage_ai/data_cleaner/column_type_detector.py:130]: This threshold seems pretty high. Do you have particular reason of choosing a high threshold?

File: mage_ai/data_cleaner/analysis/charts.py [hunk 13]
```diff
@@ -244,7 +246,8 @@ def build_overview_data(df, datetime_features):
         if df_copy[datetime_column].count() <= 1:
             continue
 
-        df_copy[datetime_column] = df[datetime_column].apply(
+        df_copy[datetime_column] = pd.to_datetime(df[datetime_column], infer_datetime_format=True, errors='coerce')
```
Comment (wangxiaoyou1993) [inline mage_ai/data_cleaner/analysis/charts.py:249]: nit: break the line

File: mage_ai/data_cleaner/shared/utils.py (hunk 11 above)
Comment (wangxiaoyou1993) [reply mage_ai/data_cleaner/shared/utils.py:28]: Yes. that make sense. To make the reformat_values rule work, we'll need to pass the uncleaned dataframe to the cleaning rule as well.

File: mage_ai/data_cleaner/column_type_detector.py (hunk 12 above)
Comment (skunichetty) [reply mage_ai/data_cleaner/column_type_detector.py:130]: My thoughts initially were that if a column is filled with a more niche data type such as emails and phone numbers, there should be a larger number of matches. That said, I can definitely think of many instances in which that wouldn't be the case and that this threshold could be too large (e.g. people fill in the email column incorrectly).

The original threshold was 0.01, which I feel to be too low. At this threshold, a large enough dataset with random strings might even accidentally match one of the patterns a few times (albeit quite rarely). 

How does a threshold of 0.3 sound? Should be more resistant to noise in the dataset yet still be flexible with regards to human error.

File: mage_ai/data_cleaner/shared/utils.py (hunk 11 above)
Comment (skunichetty) [reply mage_ai/data_cleaner/shared/utils.py:28]: I'll start working on rewriting some of the cleaning rule code so both the cleaned and uncleaned dataframes are given as inputs. In this case then, we will only be showing the uncleaned dataframe to the user, correct?

File: mage_ai/data_cleaner/column_type_detector.py (hunk 12 above)
Comment (wangxiaoyou1993) [reply mage_ai/data_cleaner/column_type_detector.py:130]: Yes, 0.3 sounds good to me.

Comment (skunichetty) [reply mage_ai/data_cleaner/column_type_detector.py]: Fixed threshold

File: mage_ai/data_cleaner/column_type_detector.py [hunk 14]
```diff
@@ -127,11 +122,11 @@ def infer_object_type(series, kwargs):
                 correct_emails = clean_series.str.match(REGEX_EMAIL).sum()
                 correct_phone_nums = clean_series.str.match(REGEX_PHONE_NUMBER).sum()
                 correct_zip_codes = clean_series.str.match(REGEX_ZIP_CODE).sum()
-                if correct_emails / length >= 0.90:
+                if correct_emails / length >= 0.30:
```
Comment (wangxiaoyou1993) [inline mage_ai/data_cleaner/column_type_detector.py:124]: nit: create constants for these threshold numbers.

File: mage_ai/data_cleaner/column_type_detector.py [hunk 15]
```diff
@@ -159,10 +154,10 @@ def infer_column_types(df, **kwargs):
     num_entries = len(df)
     if num_entries > MULTITHREAD_MAX_NUM_ENTRIES:
         types = run_parallel_multiple_args(
-            infer_column_type,
+            infer_column_type,
```
Comment (wangxiaoyou1993) [inline mage_ai/data_cleaner/column_type_detector.py:156]: nit: remove empty space after comma. same below

**Review by wangxiaoyou1993 — APPROVED** [review]
